text/0000-build-std.md

```diff
@@ -0,0 +1,4582 @@
+succeed.
+
+*See the following sections for future possibilities:*
+
+- [*Avoid building `panic_unwind` unnecessarily*][future-panic_unwind]
+
+### Special object files
+[special-object-files]: #special-object-files
+
+A handful of targets require linking against special object files, such as
+`windows-gnu`, `linux-musl` and `wasi` targets. For example, `linux-musl`
+targets require `crt1.o`, `crti.o`, `crtn.o`, etc.
+
+Since [rust#76185]/[compiler-team#343], the compiler has a stable
+`-Clink-self-contained` flag which will look for special object files in
+expected locations, typically populated by the `rust-std` components. Its
+behaviour can be forced by `-Clink-self-contained=true`, but is force-enabled
+for some targets and inferred for others.
+
+Rust can start to ship `rust-self-contained-$target` components for any targets
+which need it (including tier three targets). These components will contain the
+special object files normally included in `rust-std`, and will be distributed
+for all tiers of targets. While generally these objects are specific to the
+architecture and C runtime (CRT) (and so `rust-self-contained-$arch-$crt` could
+be sufficient and result in fewer overall components), it's technically possible
+that Rust could support two targets with the same architecture and same CRT but
+different versions of the CRT, so having target-specific components is most
+future-proof. These would replace the `self-contained` directory in existing
+`rust-std` components.
+
+As long as these components have been downloaded, as well as any other support
+components, such as `rust-mingw`, rustc's `-Clink-self-contained` will be able
+to link against the object files and build-std should never fail on account of
+missing special object files.
+
+*See the following sections for future possibilities:*
+
+- [*Enable local recompilation of special object files/sanitizer runtimes*][future-recompile-special]
+
+### `compiler-builtins-mem`
+[compiler-builtins-mem]: #compiler-builtins-mem
+
+The `mem` feature of `compiler_builtins` (and the subsequent
+`compiler-builtins-mem` feature of `core`, `alloc`, `std` which forward to
+`compiler_builtins/mem`) is required by `no_std` crates because `libc` does not
+provide these symbols without `std`.
+
+It is necessary that the `compiler-builtins-mem` feature of `alloc` and/or
+`core` be enabled when `std` is not in the crate graph
+([?][rationale-no-weak-linkage]).
+
+*See the following sections for rationale/alternatives:*
+
+- [*Why not use weak linkage for `compiler-builtins/mem` symbols?*][rationale-no-weak-linkage]
+
+### `libunwind`
+[libunwind]: #libunwind
+
+`libunwind`'s sources are included in the `rust-src` component so that they can
+be used as part of the standard library build on targets which require it.
+
+### Potential migration breakage
+[potential-migration-breakage]: #potential-migration-breakage
+
+When building an existing `no_std` project for a tier two or three target with
```

Comment:
I don't understand why this section wouldn't also apply to tier-1 targets. Also, I'm not understanding. A tier-2 target doesn't need `-Zbuild-std` to use `#![no_std]`. I don't see how those wouldn't break.

Reply:
I got slightly confused upon reading this section too. I can compile a crate that uses `#![no_std]` with `x86_64-unknown-none` (a tier-2 target) right now on stable.

After this RFC, the crate would have an implicit dependency on `std`. But `x86_64-unknown-none` has `standard_library_support.std = false`. So would cargo error?

.. or would this only apply to when cargo attempts to build std, so the implicit+loaded from sysroot path doesn't, for some reason? Should probably be specified.

Reply:
The wording here slowly evolved into something awkard - it's this way because there are no tier 1 no_std targets. Specifically this occurs if any crate in a project has an implicit dependency on std (which no_std projects may have if any crate is not updated to support builtins) and the project is built for a no_std target. I'll reword it